vertx-web/src/test/java/io/vertx/ext/web/RouterTest.java

```diff
@@ -3595,4 +3595,54 @@ public void testPausedConnection4() {
 
     await();
   }
+
+  @Test
+  public void testPathPrefixWithPathParams() throws Exception {
+    router.route("/a/:id/b/:id2/c/*").handler(rc -> {
+      assertEquals("123", rc.pathParam("id"));
+      assertEquals("1234", rc.pathParam("id2"));
+      rc.response().end("r4");
+    });
+
+    router.route("/a/:id/b1/*").handler(rc -> {
+      assertEquals("123", rc.pathParam("id"));
+      rc.response().end("r3");
+    });
+
+    router.route("/a/:id/*").handler(rc -> {
+      assertEquals("123", rc.pathParam("id"));
+      rc.response().end("r2");
+    });
+
+    router.route("/:id/*").handler(rc -> {
+      assertEquals("123", rc.pathParam("id"));
+      rc.end("r1");
+    });
+
+    // should call "/:id/*"
+    testRequest(HttpMethod.GET, "/123/foo", 200, "OK", "r1");
+    testRequest(HttpMethod.GET, "/123/", 200, "OK", "r1");
+    // should call "/a/:id/*"
+    testRequest(HttpMethod.GET, "/a//123/", 200, "OK", "r2");
+    testRequest(HttpMethod.GET, "/a/123/foo/bar", 200, "OK", "r2");
+    // should call "/a/:id/b1/*"
+    testRequest(HttpMethod.GET, "/a//123/b1/", 200, "OK", "r3");
+    testRequest(HttpMethod.GET, "/a/123/b1/foo", 200, "OK", "r3");
+    // should call "/a/:id/b/:id2/c/*"
+    testRequest(HttpMethod.GET, "/a//123/b/1234/c/", 200, "OK", "r4");
+    testRequest(HttpMethod.GET, "/a/123/b/1234/c/foo", 200, "OK", "r4");
+
+
+    // should call "/:id/*"
+    testRequest(HttpMethod.GET, "/123", 200, "OK", "r1");
+    // should call "/a/:id/*"
+    testRequest(HttpMethod.GET, "/a/123", 200, "OK", "r2");
+    testRequest(HttpMethod.GET, "/a//123", 200, "OK", "r2");
+    // should call "/a/:id/b1/*"
+    testRequest(HttpMethod.GET, "/a/123/b1", 200, "OK", "r3");
+    testRequest(HttpMethod.GET, "/a//123//b1", 200, "OK", "r3");
+    // should call "/a/:id/b/:id2/c/*"
+    testRequest(HttpMethod.GET, "/a/123/b/1234/c", 200, "OK", "r4");
+    testRequest(HttpMethod.GET, "/a//123//b//1234//c", 200, "OK", "r4");
+  }
```

Comment:
These assumptions are incorrect according to the rules the matching engine works.

A wildcard route is a route that ends with: `*` however it is important to note that the final slash is significant. At the HTTP spec level: `http://server/a` and `http://server/a/` are considered 2 different resources. In vertx-web we claim that for simplicity, a static route that doesn't include the final `/` will match both requests to the same route, however, the routes in this example end with `/*` which means that the final `/` is significative. This means the request **must** include it to be matched.

Reply:
Thanks alot for the explanation! There might still be some inconsistencies in documentation and/or behavior - see also https://github.com/vert-x3/vertx-web/issues/2395#issuecomment-1482646523